### ISSUE TITLE
[question] Define build destination 

### ISSUE THREAD

**Issue by Corallo** [conan-io/conan#13582]
### What is your question?

Hello!

I am trying to cross-build opencv (From the Conan Center) from x86 to amrv8 using Conan 1.59

To build opencv I am running the command:
`conan install . --build=missing -pr:b=default -pr:h=./cross-profile.txt -c tools.system.package_manager:mode=install -if /opt/axis/acapsdk/sysroots/aarch64/usr/lib/pkgconfig/ -of /opt/axis/acapsdk/sysroots/aarch64/usr/`

My default profile is:
```
[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]
```

I have a cross-profile.txt file that looks like this:
```
[settings]
os=Linux
arch=armv8
build_type=Release
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++11

[buildenv]
CC=aarch64-linux-gnu-gcc -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib
CXX=aarch64-linux-gnu-g++ -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib
LD=aarch64-linux-gnu-ld --sysroot=/opt/axis/acapsdk/sysroots/aarch64



[conf]
tools.build:sysroot=/opt/axis/acapsdk/sysroots/aarch64
```

and conanfile.txt:
```

[requires]
cmake/3.25.3
opencv/4.5.5

[generators]
PkgConfigDeps
MesonToolchain

[options]
opencv:with_openexr=False
opencv:with_webp=False
opencv:with_gtk=False
opencv:with_jpeg2000=False
opencv:with_v4l=False
opencv:with_ffmpeg=False
opencv:with_png=False
opencv:dnn=False

# make haders go to /opt/axis/acapsdk/sysroots/aarch64/usr/include/
# make libs go to /opt/axis/acapsdk/sysroots/aarch64/usr/lib/
[imports]
lib, *.h -> /opt/axis/acapsdk/sysroots/aarch64/usr/include/
lib, *.hpp -> /opt/axis/acapsdk/sysroots/aarch64/usr/include/
lib, *.so* -> /opt/axis/acapsdk/sysroots/aarch64/usr/lib/

```

The build succeed, looking in my installation directory (`/opt/axis/acapsdk/sysroots/aarch64/usr/lib/pkgconfig/`) specified in the install command I find all the opencv*.pc files.

Here's what I get when I cat one of them

```
cat /opt/axis/acapsdk/sysroots/aarch64/usr/lib/pkgconfig/opencv-opencv_core.pc
prefix=/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80
libdir1=${prefix}/lib
includedir1=${prefix}/include
includedir2=${prefix}/include/opencv4

Name: opencv-opencv_core
Description: Conan component: opencv-opencv_core
Version: 4.5.5
Libs: -L"${libdir1}" -lopencv_core -ltegra_hal -ldl -lm -lpthread -lrt 
Cflags: -I"${includedir1}" -I"${includedir2}" 
Requires: zlib eigen3
```
Everything seems fine, except that the prefix is set to `/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80`. I am trying to make it go instead in my sysroot so that I can easily build my application using opencv built by conan.

Now, when I try to build an application that uses opencv, I get the following error:
```
FAILED: opencv_app.p/example.cpp.o 
aarch64-linux-gnu-g++ -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib -Iopencv_app.p -I. -I.. -I/opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/zlib/1.2.13/_/_/package/115c8fa606e08a868a5ec8073c26c1a096c7263b/include -I/opt/axis/acapsdk/sysroots/aarch64/usr/include/glib-2.0 -I/opt/axis/acapsdk/sysroots/aarch64/usr/lib/glib-2.0/include -I/opt/axis/acapsdk/sysroots/aarch64/usr/include/vdo -I/opt/axis/acapsdk/sysroots/aarch64/usr/include/gio-unix-2.0 -I/opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include -I/opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include/opencv4 -I/opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/eigen/3.3.9/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/eigen3 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -O2 -pipe -g -feliminate-unused-debug-types -pthread -MD -MQ opencv_app.p/example.cpp.o -MF opencv_app.p/example.cpp.o.d -o opencv_app.p/example.cpp.o -c ../example.cpp
../example.cpp:19:10: fatal error: opencv-opencv_imgproc.hpp: No such file or directory
   19 | #include <opencv-opencv_imgproc.hpp>
```

It looks like that the headers are not found, as it looks for them in the wrong directory:
`/opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include
`
But they are in:
`/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include
`
Is there a way to make conan install the headers in the correct directory?
Possibly `/opt/axis/acapsdk/sysroots/aarch64/usr/ `

Or it has to be in that specific conan workspace?

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Corallo 

Thanks for your report.
We would need a bit more of data to understand what is happening:

- The Conan version. Very important now that Conan 2.0 is out. It seems Conan 1.X, but please confirm.
- The "build" profile. It is also necessary to use both "build" and "host" profiles for cross-compilation
- Are the ``opencv`` recipe the ones from ConanCenter or yours?
- The full output of the ``conan install`` command

Also a couple of hints:
- cmake is not a ``[requires]``, it should be in ``[tool_requires]``
- Also, to report things it would be much better if you can report with the ``conan new pkg/0.1 -m=cmake_lib`` template, as it si something way easier to reproduce, takes less time to compile, doesn't depend on possible changes on ConanCenter recipes done by the community, etc. I'd strongly suggest testing and reporting first with a simple project like that.

**Comment by Corallo** [comment]
Hello @memsharded 

- I am using Conan 1.59 
- the host profile is the one mentioned in the previous message, the cross-profile.txt, as build profile I am using "default".
Configuration for profile default:
```

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]
```
- Yes, it the standard opencv from ConanCenter
- The command is `conan install . --build=missing -pr:b=default -pr:h=./cross-profile.txt -c tools.system.package_manager:mode=install -if /opt/axis/acapsdk/sysroots/aarch64/usr/lib/pkgconfig/ -of /opt/axis/acapsdk/sysroots/aarch64/usr/ `

regarding the `conan new` suggestion my I am not trying to build my own conan package, I am trying to use the conan-built opencv to build my own application 

I am going to edit the first message to have all the info in one place

**Comment by memsharded** [comment]
I think you are misinterpreting the intent of ``-if`` and ``-of``. It is not for installing the packages themselves, the Conan packages go to the Conan cache (typically in ``<userhome>/.conan``, but you can configure it). The Conan cache has its own layout and structure of folders, cant be used for system installation of packages. If you want to extract files from Conan packages for system deployment, the recommendation is:
- In Conan 1.X, use a ``conanfile.py`` with a ``generate()`` method that does ``copy()`` iterating ``self.dependencies``
- In Conan 2.0 you can use a custom ``deployer`` to automate this task.

The ``-if, -of`` is just for the Conan generated files, like custom xxx-config.cmake to help locate the dependencies in the Conan cache.

Not saying that this is the cause of the cross-compilation error, as that would happen earlier.

The ``conan new`` is just generating a predefined simple project, that can be ``conan create .`` out of the box. Very useful for demos, and for quickly testing things. ``conan new pkg/0.1 -m=cmake_lib`` + ``conan create .`` should work out of the box.

**Comment by Corallo** [comment]
I think the cause of my issue is that my build system (meson) is looking for packages starting from my sysroot `opt/axis/acapsdk/sysroots/aarch64/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include ` even if pkg-config tells him that they are in `/root/.conan/data/opencv/4.5.5/_/_/package/9bab48b3e7c75afc311f7c109f44f30ba1b1cd80/include `.
So if I understood your suggestion, I should remove the -if and -of flags and instead of using a conan-file.txt I should use the python version and try to use it to copy the dependencies?

**Comment by memsharded** [comment]
Well, I think that is the purpose of ``PkgConfigDeps`` + ``MesonToolchain`` is let the project locate the dependencies in the Conan cache. If that is not happening, we should investigate why, trying to make it work by copying the files to that location looks like a quite problematic approach, as it will easily create conflicts and other problems, for example when managing different versions of the dependencies.

So it seems that there are 2 or 3 different things in this ticket:
- the ``opencv`` recipe from conan-center might not be cross-compiling correctly
- the consumer ``meson`` build is not locating the dependency in the Conan cache

The path to solve this would be, in order:
- First, make sure that we can do a native compile (not a cross-build yet) of the ``conan new pkg/0.1 -m=cmake_lib`` template. OpenCV is a cmake project, so this would be the closer
- When we are able to compile ``pkg/0.1``, have a simple ``meson`` project that is building and using such dependency, from the conan cache
- Try to cross compile with your configuration the above ``conan new`` template
- Try to use that cross-built ``pkg/0.1`` in the meson project
- Finally, when everything works, check what could be happening with ``opencv`` in ConanCenter, and investigate why it might not be cross-compiling correctly 

If you want to setup a small git toy repo with the contents of the meson simple project, that would be good to have a look and check things for us too.

**Comment by Corallo** [comment]
Thanks for the guidance.

I have been trying your steps, but I didn't get far. As far as I understood, you are suggesting to install a dummy package and try to build something that uses it. 
I tried first 
```
conan new pkg/0.1 -m=cmake_lib
conan create .
```
Where already I get an error
```

Installing (downloading, building) binaries...
pkg/0.1: Configuring sources in /root/.conan/data/pkg/0.1/_/_/source/.
pkg/0.1: Copying sources to build folder
pkg/0.1: Building your package in /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74
pkg/0.1: Generator txt created conanbuildinfo.txt
pkg/0.1: Calling generate()
pkg/0.1: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
pkg/0.1: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
pkg/0.1: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
pkg/0.1: Aggregating env generators
pkg/0.1: Calling build()
pkg/0.1: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/."
-- Using Conan toolchain: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++ - broken
CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/usr/bin/aarch64-linux-gnu-g++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_8b7f9/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_8b7f9.dir/build.make CMakeFiles/cmTC_8b7f9.dir/build
    gmake[1]: Entering directory '/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_8b7f9.dir/testCXXCompiler.cxx.o
    /usr/bin/aarch64-linux-gnu-g++   -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib -D_GLIBCXX_USE_CXX11_ABI=0  -O2 -pipe -g -feliminate-unused-debug-types    -m64  -fPIE -o CMakeFiles/cmTC_8b7f9.dir/testCXXCompiler.cxx.o -c /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    aarch64-linux-gnu-g++: error: unrecognized command-line option ‘-m64’
    gmake[1]: *** [CMakeFiles/cmTC_8b7f9.dir/build.make:78: CMakeFiles/cmTC_8b7f9.dir/testCXXCompiler.cxx.o] Error 1
    gmake[1]: Leaving directory '/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_8b7f9/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeOutput.log".
See also "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeError.log".
pkg/0.1: 
pkg/0.1: ERROR: Package 'd8d8aba822aaa76849d2f1bafe4a2a62a9f83b74' build failed
pkg/0.1: WARN: Build folder /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release
ERROR: pkg/0.1: Error in build() method, line 37
        cmake.configure()
        ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/."
```

It is looking for 

`"/usr/bin/aarch64-linux-gnu-g++"` I am not sure why it tries to compile for aarch64, considering that I am in a x86 container, in a clean directory. Maybe conan reads some of the environment variable (like CC / CXX ?).
Anyway, it fails because the compiler is not located there, but in my sysroot.

When I cross compile I don't get this issue, as I specify the CC in the cross-profile.txt under [buildenv] CC/CXX.


Also another note, when I cross-build opencv, the build with conan succeed, so it doesn't seem to me there's a problem with the recipe. Then, to build my application I run
```
meson setup --cross-file cross-file.ini build-test
meson compile -C build-test
```
The first command, checks for the opencv packages (asking pkg-config) and it finds it. 
(it is also possible to see it by running pkg-config --list-all. It points to the .pc file.
What gets misplaced are the headers and the .so
And maybe missplaced is not the correct word, as they end up in their conan cache location as you described.
But for some reason, meson go to look for them in a path that pre-append the sysroot.

**Comment by memsharded** [comment]
Conan default profile detection and also CMake will use CC/CXX environment variables, if you are pointing them to a different compiler, then this makes sense.
You can go and edit your default profile, point it to the system ``gcc`` to solve this.
You should be able to see the input profiles, you removed them from your logs, but ``conan create`` should be showing the configuration (will include like ``compiler.version`` that will match the cross-compiler one instead of system one). 
Calling ``conan profile new default --detect --force`` can also trigger the default profile creation again, it will report what compiler it is finding.

Please try to solve that, and lets see if we can pass the first step.

**Comment by Corallo** [comment]
Editing in the profile seems to make conan recognize the right compiler, but not cmake, that still keep using the environment CC.

```
conan create . -pr=default_native                       
Exporting package recipe
pkg/0.1 exports_sources: Copied 1 '.txt' file: CMakeLists.txt
pkg/0.1 exports_sources: Copied 1 '.cpp' file: pkg.cpp
pkg/0.1 exports_sources: Copied 1 '.h' file: pkg.h
pkg/0.1: The stored package has not changed
pkg/0.1: Exported revision: b46bf6a135f4083dfb3980781611f5cc
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=11
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
[buildenv]
CC=/usr/bin/gcc
CXX=/usr/bin/g++

pkg/0.1: Forced build from source
pkg/0.1 (test package): Installing package
Requirements
    pkg/0.1 from local cache - Cache
Packages
    pkg/0.1:d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74 - Build

Installing (downloading, building) binaries...
pkg/0.1: WARN: Build folder is dirty, removing it: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74
pkg/0.1: Copying sources to build folder
pkg/0.1: Building your package in /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74
pkg/0.1: Generator txt created conanbuildinfo.txt
pkg/0.1: Calling generate()
pkg/0.1: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
pkg/0.1: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
pkg/0.1: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
pkg/0.1: Aggregating env generators
pkg/0.1: Calling build()
pkg/0.1: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/."
-- Using Conan toolchain: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++ - broken
CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/usr/bin/aarch64-linux-gnu-g++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_dfa14/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_dfa14.dir/build.make CMakeFiles/cmTC_dfa14.dir/build
    gmake[1]: Entering directory '/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_dfa14.dir/testCXXCompiler.cxx.o
    /usr/bin/aarch64-linux-gnu-g++   -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib -D_GLIBCXX_USE_CXX11_ABI=0  -O2 -pipe -g -feliminate-unused-debug-types    -m64  -fPIE -o CMakeFiles/cmTC_dfa14.dir/testCXXCompiler.cxx.o -c /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    aarch64-linux-gnu-g++: error: unrecognized command-line option ‘-m64’
    gmake[1]: *** [CMakeFiles/cmTC_dfa14.dir/build.make:78: CMakeFiles/cmTC_dfa14.dir/testCXXCompiler.cxx.o] Error 1
    gmake[1]: Leaving directory '/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_dfa14/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeOutput.log".
See also "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/CMakeFiles/CMakeError.log".
pkg/0.1: 
pkg/0.1: ERROR: Package 'd8d8aba822aaa76849d2f1bafe4a2a62a9f83b74' build failed
pkg/0.1: WARN: Build folder /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release
ERROR: pkg/0.1: Error in build() method, line 37
        cmake.configure()
        ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/."
```

I could remove temporarily the CC variable, and put it back when reaching the cross compilation step?

**Comment by memsharded** [comment]
Sorry, that was my fault. 
That would be the default behavior in Conan 2.0, and I forgot about that is missing by default in 1.X.

In Conan 1.X you need to either:
- Define the ``[conf]`` ``tools.env.virtualenv:auto_use=True`` in your profiles or in your cache ``global.conf``, to emulate the 2.0 autogeneration of environment variables file scripts
- To explicitly declare the ``generators = "VirtualBuildEnv", "VirtualRunEnv"`` in your ``conanfile.py`` (the one generated by ``conan new`` doesn't define it, to keep the same for 1.X and 2.0)

The problem is that the environment defined by ``[buildenv]`` is not captured and used by the recipe. I have the ``tools.env.virtualenv:auto_use=True`` in my ``global.conf`` for 1.X, so I forgot about it. Please activate it and try again and let me know.
I know it needs a couple of hops to get to work, but this has been necessary to not break Conan 1.X and allowing a joint-compatible syntax with 2.0 to allow an easier upgrade to 2.0.

**Comment by Corallo** [comment]
Adding that line in the conf file, and the generators in the conanfile.py made it work, now the pkg lib is correctly built. 
```
root@7b86178816fb:/opt/app# conan create . -pr=./conan-profile-default-native.txt
WARN: Remotes registry file missing, creating default one in /root/.conan/remotes.json
Auto detecting your dev setup to initialize the default profile (/root/.conan/profiles/default)
CC and CXX: aarch64-linux-gnu-gcc  -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib, aarch64-linux-gnu-g++  -mcpu=cortex-a53 -march=armv8-a+crc+crypto -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/opt/axis/acapsdk/sysroots/aarch64 -L/opt/axis/acapsdk/sysroots/aarch64/usr/lib 
Found gcc 11
gcc>=5, using the major as version

************************* WARNING: GCC OLD ABI COMPATIBILITY ***********************
 
Conan detected a GCC version > 5 but has adjusted the 'compiler.libcxx' setting to
'libstdc++' for backwards compatibility.
Your compiler is likely using the new CXX11 ABI by default (libstdc++11).

If you want Conan to use the new ABI for the default profile, run:

    $ conan profile update settings.compiler.libcxx=libstdc++11 default

Or edit '/root/.conan/profiles/default' and set compiler.libcxx=libstdc++11

************************************************************************************



Default settings
        os=Linux
        os_build=Linux
        arch=x86_64
        arch_build=x86_64
        compiler=gcc
        compiler.version=11
        compiler.libcxx=libstdc++
        build_type=Release
*** You can change them in /root/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***


Exporting package recipe
pkg/0.1 exports_sources: Copied 1 '.txt' file: CMakeLists.txt
pkg/0.1 exports_sources: Copied 1 '.cpp' file: pkg.cpp
pkg/0.1 exports_sources: Copied 1 '.h' file: pkg.h
pkg/0.1: A new conanfile.py version was exported
pkg/0.1: Folder: /root/.conan/data/pkg/0.1/_/_/export
pkg/0.1: Exported revision: 5349acae6e1ae6eddfdc062b9096db34
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=11
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
[conf]
tools.env.virtualenv:auto_use=True
[buildenv]
CC=/usr/bin/gcc
CXX=/usr/bin/g++

pkg/0.1: Forced build from source
pkg/0.1 (test package): Installing package
Requirements
    pkg/0.1 from local cache - Cache
Packages
    pkg/0.1:d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74 - Build

Installing (downloading, building) binaries...
pkg/0.1: Configuring sources in /root/.conan/data/pkg/0.1/_/_/source/.
pkg/0.1: Copying sources to build folder
pkg/0.1: Building your package in /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74
pkg/0.1: Generator 'VirtualBuildEnv' calling 'generate()'
pkg/0.1: Generator 'VirtualRunEnv' calling 'generate()'
pkg/0.1: Calling generate()
pkg/0.1: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
pkg/0.1: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
pkg/0.1: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
pkg/0.1: Aggregating env generators
pkg/0.1: Calling build()
pkg/0.1: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/."
-- Using Conan toolchain: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


-- Build files have been written to: /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release
pkg/0.1: CMake command: cmake --build "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release" '--' '-j16'
[ 50%] Building CXX object CMakeFiles/pkg.dir/src/pkg.cpp.o
[100%] Linking CXX static library libpkg.a
[100%] Built target pkg
pkg/0.1: Package 'd8d8aba822aaa76849d2f1bafe4a2a62a9f83b74' built
pkg/0.1: Build folder /root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release
pkg/0.1: Generated conaninfo.txt
pkg/0.1: Generated conanbuildinfo.txt
pkg/0.1: Generating the package
pkg/0.1: Package folder /root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74
pkg/0.1: Calling package()
pkg/0.1: CMake command: cmake --install "/root/.conan/data/pkg/0.1/_/_/build/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/build/Release" --prefix "/root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74"
-- Install configuration: "Release"
-- Installing: /root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/lib/libpkg.a
-- Installing: /root/.conan/data/pkg/0.1/_/_/package/d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74/include/pkg.h
pkg/0.1 package(): Packaged 1 '.a' file: libpkg.a
pkg/0.1 package(): Packaged 1 '.h' file: pkg.h
pkg/0.1: Package 'd8d8aba822aaa76849d2f1bafe4a2a62a9f83b74' created
pkg/0.1: Created package revision 60b05acffad0b3a755a7957e04e16a59
pkg/0.1 (test package): Generator txt created conanbuildinfo.txt
pkg/0.1 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
pkg/0.1 (test package): Generator 'CMakeToolchain' calling 'generate()'
pkg/0.1 (test package): Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
pkg/0.1 (test package): If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/opt/app/test_package/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
pkg/0.1 (test package): Generator 'VirtualBuildEnv' calling 'generate()'
pkg/0.1 (test package): Generator 'VirtualRunEnv' calling 'generate()'
pkg/0.1 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
pkg/0.1 (test package): Generator 'CMakeDeps' calling 'generate()'
pkg/0.1 (test package): Aggregating env generators
pkg/0.1 (test package): Generated conaninfo.txt
pkg/0.1 (test package): Generated graphinfo
Using lockfile: '/opt/app/test_package/build/Release/generators/conan.lock'
Using cached profile from lockfile
pkg/0.1 (test package): Calling build()
pkg/0.1 (test package): CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/opt/app/test_package/build/Release/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/opt/app/test_package/."
-- Using Conan toolchain: /opt/app/test_package/build/Release/generators/conan_toolchain.cmake
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Target declared 'pkg::pkg'
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


-- Build files have been written to: /opt/app/test_package/build/Release
pkg/0.1 (test package): CMake command: cmake --build "/opt/app/test_package/build/Release" '--' '-j16'
[ 50%] Building CXX object CMakeFiles/example.dir/src/example.cpp.o
[100%] Linking CXX executable example
[100%] Built target example
pkg/0.1 (test package): Running test()
pkg/0.1: Hello World Release!
  pkg/0.1: __x86_64__ defined
  pkg/0.1: _GLIBCXX_USE_CXX11_ABI 0
  pkg/0.1: __cplusplus201703
  pkg/0.1: __GNUC__11
  pkg/0.1: __GNUC_MINOR__3
```

I don't see it however when I run pkg-config --list-all, and but building an application with meson using that pkg fails, I think it is because also meson is reading the evirorment CC and CXX, and tries to use the cross-compiler
```
#include <pkg>

using namespace cv;

int main(int argc, char* argv[]) {
  printf("Hello world")
  return 0;
}
```
`meson setup  build-test`

`meson.build:1:0: ERROR: Could not invoke sanity test executable: [Errno 8] Exec format error: '/opt/app/build-test/meson-private/sanitycheckcpp.exe'.`

**Comment by memsharded** [comment]
Yes, it is expected that ``pkg-config --list-all``, as this package is not installed system-wide, it is installed in the Conan cache.

The next step is to consume it with something like:
```
[requires]
pkg/0.1

[generators]
PkgConfigDeps
MesonToolchain
VirtualBuildEnv
VirtualRunEnv
```
Then ``conan install .``, and then calling Meson passing the conan generated files

Some notes:
- I see that you are running in ``/opt``. This sounds weird, it is likely that this won't be the recommended practice. Instead, Conan should be typically executed in user folders like ``/home/myuser``
- It seems that you went directly for the cross-compilation. I'd still recommend to start with plain native compilation, until getting familiarized with the tooling and the integrations, and only then moving to cross-building
- When cross building you were still using the legacy way without providing the "build" profile, you did ``conan create . -pr=./conan-profile-default-native.txt``. For cross building it is expected to pass also the "build" profile, like ``conan create . -pr:b=default -pr=:/conan-profile-default-native.txt``

**Comment by Corallo** [comment]
Consuming the package also worked.

```
conan install .
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=11
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.txt: Installing package
Requirements
    pkg/0.1 from local cache - Cache
Packages
    pkg/0.1:d8d8aba822aaa76849d2f1bafe4a2a62a9f83b74 - Cache

Installing (downloading, building) binaries...
pkg/0.1: Already installed!
conanfile.txt: Generator 'VirtualBuildEnv' calling 'generate()'
conanfile.txt: Generator 'PkgConfigDeps' calling 'generate()'
conanfile.txt: Generator 'VirtualRunEnv' calling 'generate()'
conanfile.txt: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
conanfile.txt: Generator 'MesonToolchain' calling 'generate()'
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo
```
Regarding the extra notes, it was not my intention to do cross-compilation, I was just trying to compile using meson and referincing that lib, to familiarize with the tool, somehow meson tried to crosscompile anyway (I guess for env variables.) 
But in any case, building the package that depends on pkg seemed to work with conan without using meson directly.

Now that I know that this work, should I try to run the same, but cross-compiling? (giving the -pr:b and -pr:h flags?)

**Comment by memsharded** [comment]
I am missing the actual Meson build after the consumption with ``conan install``, did that also work? It would be great to have the output of that too.

If it worked, sure, lets move to cross-compilation, first creating the ``pkg/0.1`` with the 2 profiles, then installing it for the meson consumer with the 2 profiles, finally building it with meson.

**Comment by Corallo** [comment]
Right, I think I misunderstood something then.

`conan new pkg/0.1 -m=cmake_lib`
creates a new package that is called pkg/0.1
`conan create . -pr=./conan-profile-default-native.txt`

(Note, conan-profile-default-native.txt is just to avoid conan to use its own default, that gets confuse because of the environment variables CC and CXX) Here's the content, it is not to cross compile.
```

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
tools.env.virtualenv:auto_use=True
[build_requires]
[env]
[buildenv]
CC=/usr/bin/gcc
CXX=/usr/bin/g++

```

builds it and store it in the conan cache
Not sure then what is the point of  `conan install . ` in this case.

Anyway trying to build with meson fails at this point

```
CC=gcc CXX=g++ meson setup build-test
The Meson build system
Version: 1.0.1
Source dir: /opt/app
Build dir: /opt/app/build-test
Build type: native build
Project name: test_app
Project version: undefined
C++ compiler for the host machine: g++ (gcc 11.3.0 "g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C++ linker for the host machine: g++ ld.bfd 2.38
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Found CMake: /usr/bin/cmake (3.22.1)
Run-time dependency pkg found: NO (tried pkgconfig and cmake)

meson.build:3:0: ERROR: Dependency "pkg" not found, tried pkgconfig and cmake

A full log can be found at /opt/app/build-test/meson-logs/meson-log.txt
```
The mesonfile is
```
project('test_app', 'cpp')

gtkdeps = [dependency('pkg')]

executable('test_app', 'example.cpp', dependencies: gtkdeps )

```
and the example.cpp

```

#include <pkg>

using namespace cv;

int main(int argc, char* argv[]) {
  printf("Hello world")
  return 0;
}

```


The log mentioned in the meson output contains:
```

cat /opt/app/build-test/meson-logs/meson-log.txt
Build started at 2023-04-03T12:25:24.657351
Main binary: /usr/bin/python3
Build Options: 
Python system: Linux
The Meson build system
Version: 1.0.1
Source dir: /opt/app
Build dir: /opt/app/build-test
Build type: native build
Project name: test_app
Project version: undefined
-----
Detecting compiler via: g++ --version
compiler returned <Popen: returncode: 0 args: ['g++', '--version']>
compiler stdout:
g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


compiler stderr:

Running command: g++ -E -dM -
-----
Detecting linker via: g++ -Wl,--version -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed -Wl,-z,relro,-z,now -Wl,--disable-new-dtags -O2 -pipe -g -feliminate-unused-debug-types
linker returned <Popen: returncode: 0 args: ['g++', '-Wl,--version', '-Wl,-O1', '-Wl,--hash-...>
linker stdout:
GNU ld (GNU Binutils for Ubuntu) 2.38
Copyright (C) 2022 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.

linker stderr:
collect2 version 11.3.0
/usr/bin/ld -plugin /usr/lib/gcc/x86_64-linux-gnu/11/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper -plugin-opt=-fresolution=/tmp/ccD8J1Gt.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -z now -z relro /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/Scrt1.o /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/11/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/11 -L/usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/11/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/11/../../.. --version -O1 --hash-style=gnu --as-needed -z relro -z now --disable-new-dtags -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/11/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crtn.o

Sanity testing C++ compiler: g++
Is cross compiler: False.
Sanity check compiler command line: g++ sanitycheckcpp.cc -o sanitycheckcpp.exe -O2 -pipe -g -feliminate-unused-debug-types -D_FILE_OFFSET_BITS=64 -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed -Wl,-z,relro,-z,now -Wl,--disable-new-dtags
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command:  /opt/app/build-test/meson-private/sanitycheckcpp.exe
C++ compiler for the host machine: g++ (gcc 11.3.0 "g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C++ linker for the host machine: g++ ld.bfd 2.38
-----
Detecting compiler via: g++ --version
compiler returned <Popen: returncode: 0 args: ['g++', '--version']>
compiler stdout:
g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


compiler stderr:

Running command: g++ -E -dM -
-----
Detecting linker via: g++ -Wl,--version -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed -Wl,-z,relro,-z,now -Wl,--disable-new-dtags -O2 -pipe -g -feliminate-unused-debug-types
linker returned <Popen: returncode: 0 args: ['g++', '-Wl,--version', '-Wl,-O1', '-Wl,--hash-...>
linker stdout:
GNU ld (GNU Binutils for Ubuntu) 2.38
Copyright (C) 2022 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.

linker stderr:
collect2 version 11.3.0
/usr/bin/ld -plugin /usr/lib/gcc/x86_64-linux-gnu/11/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper -plugin-opt=-fresolution=/tmp/ccLo03wm.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -z now -z relro /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/Scrt1.o /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/11/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/11 -L/usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/11/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/11/../../.. --version -O1 --hash-style=gnu --as-needed -z relro -z now --disable-new-dtags -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/11/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crtn.o

Sanity testing C++ compiler: g++
Is cross compiler: False.
Sanity check compiler command line: g++ sanitycheckcpp.cc -o sanitycheckcpp.exe -O2 -pipe -g -feliminate-unused-debug-types -D_FILE_OFFSET_BITS=64 -Wl,-O1 -Wl,--hash-style=gnu -Wl,--as-needed -Wl,-z,relro,-z,now -Wl,--disable-new-dtags
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command:  /opt/app/build-test/meson-private/sanitycheckcpp.exe
C++ compiler for the build machine: g++ (gcc 11.3.0 "g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C++ linker for the build machine: g++ ld.bfd 2.38
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Pkg-config binary for 1 is not cached.
Pkg-config binary for 1 specified from cross file, native file, or env var as ['pkg-config']
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Determining dependency 'pkg' with pkg-config executable '/usr/bin/pkg-config'
env[PKG_CONFIG_PATH]: /opt/axis/acapsdk/sysroots/aarch64/usr/lib/pkgconfig:/opt/axis/acapsdk/sysroots/aarch64/usr/share/pkgconfig
env[PKG_CONFIG_SYSROOT_DIR]: /opt/axis/acapsdk/sysroots/aarch64
env[PKG_CONFIG]: pkg-config
Called `/usr/bin/pkg-config --modversion pkg` -> 1
stderr:
Package pkg was not found in the pkg-config search path.
Perhaps you should add the directory containing `pkg.pc'
to the PKG_CONFIG_PATH environment variable
No package 'pkg' found
-----------
CMake binary for 1 is not cached
CMake binary missing from cross or native file, or env var undefined.
Trying a default CMake fallback at cmake
Found CMake: /usr/bin/cmake (3.22.1)
Extracting basic cmake information
CMake Toolchain: Calling CMake once to generate the compiler state
Calling CMake (['/usr/bin/cmake']) in /opt/app/build-test/meson-private/__CMake_compiler_info__ with:
  - "--trace-expand"
  - "--trace-format=json-v1"
  - "--no-warn-unused-cli"
  - "--trace-redirect=cmake_trace.txt"
  - "-G"
  - "Ninja"
  - "-DCMAKE_TOOLCHAIN_FILE=/opt/app/build-test/meson-private/__CMake_compiler_info__/CMakeMesonTempToolchainFile.cmake"
  - "."
Try CMake generator: auto
Calling CMake (['/usr/bin/cmake']) in /opt/app/build-test/meson-private/cmake_pkg with:
  - "--trace-expand"
  - "--trace-format=json-v1"
  - "--no-warn-unused-cli"
  - "--trace-redirect=cmake_trace.txt"
  - "-DCMAKE_TOOLCHAIN_FILE=/opt/app/build-test/meson-private/cmake_pkg/CMakeMesonToolchainFile.cmake"
  - "."
  -- Module search paths:    ['/', '/opt', '/usr', '/usr/local']
  -- CMake root:             /usr/share/cmake-3.22
  -- CMake architectures:    ['aarch64-linux-gnu', 'x86_64-linux-gnu']
  -- CMake lib search paths: ['lib', 'lib32', 'lib64', 'libx32', 'share', 'lib/aarch64-linux-gnu', 'lib/x86_64-linux-gnu']
Preliminary CMake check failed. Aborting.
Run-time dependency pkg found: NO (tried pkgconfig and cmake)

meson.build:3:0: ERROR: Dependency "pkg" not found, tried pkgconfig and cmake
```

**Comment by memsharded** [comment]
Hi @Corallo 

Sorry, this thread fell through the cracks, I couldn't follow it.

It seems that I am still missing one piece, after the dependency ``pkg/0.1`` is created and before you call meson in ``CC=gcc CXX=g++ meson setup build-test``, I am missing a ``conan install`` for the meson project, to install the meson project dependencies (in this case, the ``pkg/0.1`` dependency, defined in a ``conanfile.txt``). Maybe you are conflating the ``conanfile`` of the ``pkg/0.1`` with the necessary one for the consumer meson project?

Then also, the meson would be missing some arguments in command line to pass the toolchain files generated by ``MesonToolchain`` and ``PkgConfigDeps``, it seems you are not providing them, so it is expected that meson won't be able to find the ``pkg/0.1`` dependency, as nobody is passing the information to it.

If that doesn't help, the best is to setup a github repo with a simple ``test.py`` script that executes the different steps, like ``conan new`` + ``conan create . -pr=profile`` + ``cd ../mymeson_project`` + ``conan install .`` + ``meson ....``, and the necessary code to show the case (everything can be simple ``hello.cpp`` programs.

**Comment by Corallo** [comment]
Hey @memsharded 
In the menatime I have digged into it and I found out this:
In my envirorment I have set the variable:
`PKG_CONFIG_SYSROOT_DIR=/opt/axis/acapsdk/sysroots/aarch64`
reading [here](https://autotools.info/pkgconfig/cross-compiling.html), this variable is used to inject the sysroot when cross compiling between the -I/-L and the package location.
This is done to make sure that the cross compilation environment is not mixed up with the environment of the building container.

For example, when building with meason, **without** that flag, it adds `-I/usr/include/glib-2.0` to the compiler command which is incorrect, as that is the user/include lib of the container, and not of the target device.
So, to correctly cross compile, adding that env variable, the sysroot is injected, becoming `-I/opt/axis/acapsdk/sysroots/aarch64/usr/include/glib-2.0`.

So far so good, the problem is that this injection happens also for conan packages, so from `-I/root/.conan2/p/...` I get `-I/opt/axis/acapsdk/sysroots/aarch64/root/.conan2/p/...` which in this case is not desired.

For now I am just copying the conan cache, but I'm looking for a better solution.
Do you have any suggestions?

**Comment by memsharded** [comment]
Hi @Corallo 

Thanks for following up. 
I am afraid that this issue is very complicated to help. It has too many moving pieces, it would be very challenging to reproduce to understand what is happening and how could it be improved. I am not sure that setting PKG_CONFIG_SYSROOT_DIR is the way to go, because if it is something that Conan or Meson are not aware of, then it is very difficult or impossible to produce the right build for it.

I still think it would be necessary to setup something that can be reproduced, like a small project, in 2 steps:
- First the very basic project, that does a native build of a Meson project with 1 very simple "hello world" dependency (the dependency could use CMake if this simplifies something)
- Then, the same project, with a docker image that contains the cross compiling, and then by defining new profiles, it should be possible to cross-build the same project.

As I understand it, it shouldn't be necessary to do a copy of the dependency to the sysroot, in fact, it sounds like a problematic approach and it should be avoided. But we need to understand better the problem and the constraints.

**Comment by memsharded** [comment]
Closing this ticket as inactive and outdated (using legacy Conan 1, it might be very possible that Conan 2 doesn't have the same issues).

Please create a new ticket against Conan 2 if necessary and this is still an issue, if so, it would be great to have a reproducible example to investigate, as suggested above. Thanks very much for your feedback.
